tests/hazmat/primitives/test_ec.py

```diff
@@ -749,3 +765,99 @@ def test_signature_not_bytes(self, backend):
+class NullKDF(object):
+    def derive(self, key_material):
+        return key_material
+
+    def verify(self, key_material, expected_key):
+        if not constant_time.bytes_eq(key_material, expected_key):
+            raise exceptions.InvalidKey
+
+
+@pytest.mark.requires_backend_interface(interface=EllipticCurveBackend)
+class TestECDHVectors(object):
+
+    def test_unsupported_exchange_algorithm(self, backend):
+        curve = ec.SECP521R1()
+        exchange_alg = DummyKeyExchange()
+        _skip_exchange_algorithm_unsupported(backend, exchange_alg, curve)
+
+    def test_invalid_exchange_algorithm(self, backend):
+        curve = ec.SECP521R1
+        _skip_curve_unsupported(backend, curve)
+        key = ec.generate_private_key(curve, backend)
+        with pytest.raises(exceptions.UnsupportedAlgorithm):
+            key.exchange(DummyKeyExchange())
+
+    def test_unsupported_key_derivation_function(self, backend):
+        curve = ec.SECP521R1
+        _skip_curve_unsupported(backend, curve)
+        key = ec.generate_private_key(curve, backend)
+        with pytest.raises(TypeError):
+            key.exchange(ec.ECDH(DummyKeyExchange()))
+
+    def test_invalid_key_verification(self, backend):
+        with pytest.raises(exceptions.InvalidKey):
+            NullKDF().verify(b'a', b'b')
+        NullKDF().verify(b'a', b'a')
+
+    def key_exchange(self, backend, vector):
+        key_numbers = vector['IUT']
+        peer_numbers = vector['CAVS']
+
+        key = ec.EllipticCurvePrivateNumbers(
+            key_numbers['d'],
+            ec.EllipticCurvePublicNumbers(
+                key_numbers['x'],
+                key_numbers['y'],
+                ec._CURVE_TYPES[vector['curve']]()
+            )
+        ).private_key(backend)
+
+        peerkey = ec.EllipticCurvePrivateNumbers(
+            peer_numbers['d'],
+            ec.EllipticCurvePublicNumbers(
+                peer_numbers['x'],
+                peer_numbers['y'],
+                ec._CURVE_TYPES[vector['curve']]()
+            )
+        ).private_key(backend)
+        pubkey = peerkey.public_key()
+
+        keyex = key.exchange(ec.ECDH(NullKDF()))
+        z = keyex.agree(pubkey)
+
+        return int(hexlify(z).decode('ascii'), 16)
+
+    @pytest.mark.parametrize(
+        "vector",
+        load_vectors_from_file(
+            os.path.join(
+                "asymmetric", "ECDH",
+                "KASValidityTest_ECCStaticUnified_NOKC_ZZOnly_init.fax"),
+            load_kasvs_ecdh_vectors
+        )
+    )
+    def test_key_exchange_with_vectors(self, backend, vector):
+        _skip_curve_unsupported(backend, ec._CURVE_TYPES[vector['curve']])
+        _skip_exchange_algorithm_unsupported(backend, ec.ECDH(NullKDF()),
+                                             ec._CURVE_TYPES[vector['curve']])
+
+        try:
+            z = self.key_exchange(backend, vector)
+        except ValueError:
+            assert vector['fail'] is True
```

Comment:
Maybe a silly question, but why do the vectors have cases that fail before
we ever get to doing ECDH with them?

On Wed, Oct 14, 2015 at 11:27 PM, simo5 <notifications@github.com> wrote:

> In tests/hazmat/primitives/test_ec.py
> <https://github.com/pyca/cryptography/pull/2398#discussion_r42080246>:
>
> > +        load_vectors_from_file(
> > +            os.path.join(
> > +                "asymmetric", "ECDH",
> > +                "KASValidityTest_ECCStaticUnified_NOKC_ZZOnly_init.fax"),
> > +            load_kasvs_ecdh_vectors
> > +        )
> > +    )
> > +    def test_key_exchange_with_vectors(self, backend, vector):
> > +        _skip_curve_unsupported(backend, ec._CURVE_TYPES[vector['curve']])
> > +        _skip_exchange_algorithm_unsupported(backend, ec.ECDH(NullKDF()),
> > +                                             ec._CURVE_TYPES[vector['curve']])
> > +
> > +        try:
> > +            z = self.key_exchange(backend, vector)
> > +        except ValueError:
> > +            assert vector['fail'] is True
>
> generally turning numbers into keys
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/2398/files#r42080246>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084